File: hasekll/Calc/ToRPN.hs
module ToRPN (toRPN) where

import Types
import Data.Ord (comparing)

toRPN :: [Token] -> [Token]
toRPN tokens = reverse $ toRPN' tokens [] []
  where
    toRPN' [] stack output = popAll stack output
    toRPN' toekns@(t:ts) stack output =
        case t of
          Var _          -> toRPN' ts stack (t : output)
          Const _        -> toRPN' ts stack (t : output)
          LeftParen      -> toRPN' ts (t : stack) output
          RightParen     -> let (stack', output')
                                    = popUntilLeftParen stack output
                            in toRPN' ts stack' output'
          Operator _ _ _ -> let (stack', output')
                                    = popOperator t stack output
                            in toRPN' ts (t : stack') output'

    popAll [] output = output
    popAll tokens@(t:ts) output =
        case t of
          Operator _ _ _ -> popAll ts (t : output)
          illegalToken   -> error "illegal token lefted in stack"

    popUntilLeftParen [] output = error "paren not match"
    popUntilLeftParen (top:xs) output =
        case top of
          LeftParen -> (xs, output)
          token     -> popUntilLeftParen xs (top : output)

    popOperator op [] output = ([], output)
    popOperator op stack@(top:xs) output
        | isOperator top && shouldPop = popOperator op xs (top : output)
        | otherwise                   = (stack, output)
      where
        shouldPop
            -- if current operator(op) is unary, don't pop stack
            | unAssoc op == None = False
            -- if top of stack is unary
            -- pop only if op's precedence < top's (TODO: < or <=)
            | unAssoc top == None && assocCompare == LT = True
            -- neighter top or currnt operator is unary
            -- pop if top is left  assoc and op's precedence <= top's
            -- or     top is right assoc and op's precedence == top's
            | otherwise =
                or [ unAssoc top == LeftAssoc  && assocCompare /= GT
                   , unAssoc top == RightAssoc && assocCompare == LT
                   ]
        assocCompare = comparing unPred op top

File: c/tautology/to_rpn.c
#include <stdio.h>
#include "to_rpn.h"
#include "token.h"
#include "stack.h"
#include "expression.h"

void pop_all(struct stack *stack, struct stack *output)
{
	while (!is_empty(stack)) {
		push(output, pop(stack));
	}
}

void pop_until_leftparen(struct stack *stack, struct stack *output)
{
	struct token * token = get_top(stack);
	if (token->type == LeftParen) {
		return;
	}
	push(output, pop(stack));
	pop_until_leftparen(stack, output);
}

int should_pop(struct token *token,
	       struct stack *stack,
	       struct stack *output)
{
	if (token->type == OpNot) {
		return False;
	}
	struct token *top = get_top(stack);
	if (top->type == OpNot) {
		return True;
	}
	if (token->type - top->type <= 0) {
		return True;
	}
	return False;
}


void pop_op(struct token *token,
	    struct stack *stack,
	    struct stack *output)
{
	if (token->type < OpOr || !should_pop(token, stack, output)) {
		return;
	}
	push(output, pop(stack));
	pop_op(++token, stack, output);
}


bool to_rpn(struct token *tokens,
       struct stack *stack,
       struct stack *output)
{

	struct token *curr;
	while (tokens->type == NotToken) {
		curr = tokens;
		switch (curr->type) {
		case Var:
			push(output, new_var_expression(curr));
			break;
		case LeftParen:

			push(output, new_paren_expression(curr));
			break;
		case RightParen:
			pop_until_leftparen(stack, output);
			break;
		case OpOr:
		case OpAnd:
		case OpNot:
			pop_op(curr, stack, output);
			push(stack, new_op_expression(curr));
		default:
			return False;
		}
		tokens++;
	}
	pop_all(stack, output);

	return True;
}
